### PR TITLE
Make file upload notification callback async and update doc comments

### PR DESCRIPTION
**Pull request by drwill-ms**

### CODE REVIEW

**Review by brycewang-microsoft — APPROVED** [review]

**Review by patilsnr — APPROVED** [review]

File: e2e/Tests/iothub/service/FileUploadNotificationE2ETest.cs [hunk 1]
```diff
@@ -27,13 +31,15 @@ public class FileUploadNotificationE2ETest : E2EMsTestBase
...
         // the service makes it available for redelivery to other open receivers.
-        private readonly AcknowledgementType _acknowledgementType = AcknowledgementType.Abandon;
+        private readonly AcknowledgementType _defaultAcknowledgementType = AcknowledgementType.Abandon;
 
         [TestMethod]
-        [Timeout(LongRunningTestTimeoutMilliseconds)]
-        [DataRow(IotHubTransportProtocol.Tcp)]
-        [DataRow(IotHubTransportProtocol.WebSocket)]
-        public async Task FileUploadNotification_Operation(IotHubTransportProtocol protocol)
+        [DataRow(IotHubTransportProtocol.Tcp, 1, false)]
```
Comment (drwill-ms) [inline e2e/Tests/iothub/service/FileUploadNotificationE2ETest.cs:37]: 3 tests consolidated to 1 data-driven test.

File: e2e/Tests/iothub/service/FileUploadNotificationE2ETest.cs [hunk 2]
```diff
@@ -42,59 +48,65 @@ public async Task FileUploadNotification_Operation(IotHubTransportProtocol proto
...
 
-            var counter = new FileUploadNotificationCounter();
-            Func<FileUploadNotification, AcknowledgementType> OnFileUploadNotificationReceived = (fileUploadNotification) =>
+            try
             {
-                counter.FileUploadNotificationsReceived++;
-                return _acknowledgementType;
-            };
+                var files = new Dictionary<string, bool>(filesToUpload);
+                var allFilesFound = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```
Comment (drwill-ms) [inline e2e/Tests/iothub/service/FileUploadNotificationE2ETest.cs:54]: Use a task completion source to know when the callback has all it needs.

File: e2e/Tests/iothub/service/FileUploadNotificationE2ETest.cs [hunk 3]
```diff
@@ -42,59 +48,65 @@ public async Task FileUploadNotification_Operation(IotHubTransportProtocol proto
...
+                await serviceClient.FileUploadNotifications.CloseAsync().ConfigureAwait(false);
+            }
         }
 
         [TestMethod]
-        [Timeout(TestTimeoutMilliseconds)]
         [DataRow(IotHubTransportProtocol.Tcp)]
         [DataRow(IotHubTransportProtocol.WebSocket)]
-        public async Task FileUploadNotification_ReceiveMultipleNotificationsInOneConnection(IotHubTransportProtocol protocol)
+        public async Task FileUploadNotification_ErrorProcessor_ReceivesNotifications(IotHubTransportProtocol protocol)
```
Comment (drwill-ms) [inline e2e/Tests/iothub/service/FileUploadNotificationE2ETest.cs:109]: Add a test for ErrorProcessor

**Pull request merged**